pkg/util/utils.go

```diff
@@ -79,3 +91,73 @@ func CommaStringToStringMap(s string) (map[string]string, error) {
 
 	return kvMap, nil
 }
+
+// ValidateIpFormat will check whether the string is in IP address format(#.#.#.#).
+func ValidateIpFormat(ip string) bool {
+	r := regexp.MustCompile(`\b(\d+)\.(\d+)\.(\d+)\.(\d+)\b`)
+	if r.MatchString(ip) {
+		return true
+	} else {
+		return false
+	}
+}
+
+// ValidateEndpoint will valid whether given endpoint is a valid.
+// Following validation will be done on the endpoint:
+// 1. Check whether the host part (IP or hostname) is valid
+// 2. Check whether port number is present in the endpoint, If not add Default_port.
+func ValidateEndpoint(endpoint string) (string, error) {
+	ip := net.ParseIP(endpoint)
+	if ip == nil {
+		host, port, _ := net.SplitHostPort(endpoint)
+		if len(port) == 0 && len(host) != 0 {
+			// Missing port after ":", add the DefaultPort value
+			// For example: "localhost:" or "192.168.110.0:"
+			if govalidator.IsIP(host) || govalidator.IsHost(host) {
+				updatedEndpoint := net.JoinHostPort(host, DefaultPort)
+				return updatedEndpoint, nil
+			} else {
+				return "", ErrInvalidEndpoint
+			}
+		} else if len(port) != 0 && len(host) != 0 {
+			// Endpoint has both host and port. Is it valid?
+			// For example:
+			// Valid: "192.168.1.1:9020" or "localhost:9020"
+			// InValid: "192.168.1.500:9020" or "local*host:9020"
+			if ValidateIpFormat(host) {
```

Comment:
if use of govalidator.IsIPv4() then this can be made easier.

Reply:
Basically,  I used the validateIPFormat and then IsIP because, I would like to classify invalid IP address as invalid IP address instead of Valid Hostname. More details below:

```
package main

import (
    “fmt”
    “regexp”
    “github.com/asaskevich/govalidator”
)

// ValidateIpFormat will check whether the string is in IP address format(#.#.#.#).
func ValidateIpFormat(ip string) bool {
    r := regexp.MustCompile(`\b(\d+)\.(\d+)\.(\d+)\.(\d+)\b`)
    if r.MatchString(ip) {
        return true
    } else {
        return false
    }
}

func main() {
    ip1 := “1.1.1.1"
    fmt.Println(“IP1 status: “, govalidator.IsIPv4(ip1))
    ip2 := “500.500.500.500”
    fmt.Println(“IP2 status: “, govalidator.IsIPv4(ip2))
    
    //Case 1
    if govalidator.IsIPv4(ip2) {
        fmt.Println(“Case1: Valid Ip address”)
    } else if govalidator.IsHost(ip2) {
        fmt.Println(“Case1: Valid Host”)
    } else {
        fmt.Println(“Case1: Invalid endpoint”)
    }
    
    // Case2:
    if ValidateIpFormat(ip2) {
        if govalidator. IsIPv4(ip2) {
            fmt.Println(“Case2: Valid Ip address”)
        } else {
            fmt.Println(“Case2: Invalid Ip address”)
        }
    } else {
        if govalidator.IsHost(ip2) {
            fmt.Println(“Case2: Valid Host name”)
        } else {
            fmt.Println(“Case2: Invalid Host name”)
        }
    }
}

```
```
Output:
IP1 status:  true
IP2 status:  false
Case1: Valid Host
Case2: Invalid Ip address
```


Here when i send the 500.500.500.500 as IP address value to govalidator.IsIPv4 or govalidator.IsIP, it will say directly that it is not a valid IP address and if I pass the same string to govalidatoe.IsHost(), it will say, it is a valid hostname.

So first I check whether, we have string of valid IP address format then use IsIP address to validate it.
If it is not of valid IP address format (#.#.#.#), I will check for the valid host name format.

That is why I used validateIpFormat first and used IsIP or IsHost() in the logic.
Please let me know, if it clarifies the queries.